### PR TITLE
Migrate sensitivities of solution and value function of parameters to `p_global`

### PR DESCRIPTION
**Pull request by FreyJo**
Solution sensitivities with respect to parameters and derivatives of the value function with respect to parameters are now computed with respect to `p_global` instead of `p`. This is anyway what was implemented conceptually. However, if different parameter values for stages were used, these sensitivities were not clearly defined.
The precomputations and the fact that the parameter vector is only defined once makes this implementation more efficient.

- added support for `p_global` in Cython

Breaking:
- concerns options `with_solution_sens_wrt_params`, `with_solution_sens_wrt_params`
- `eval_solution_sensitivity` with argument `with_respect_to = "params_global"`
- `eval_and_get_optimal_value_gradient` with argument `with_respect_to = "params_global"`
- rename option `with_respect_to='params_global'` to `'p_global'` with deprecation warning

### CODE REVIEW

File: acados/ocp_nlp/ocp_nlp_common.c [hunk 1]
```diff
@@ -3219,29 +3239,29 @@ void ocp_nlp_common_eval_lagr_grad_p(ocp_nlp_config *config, ocp_nlp_dims *dims,
     else if (!strcmp("params_global", field))
```
Comment (sandmaennchen) [inline acados/ocp_nlp/ocp_nlp_common.c:3239]: rename to `p_global`?

File: acados/ocp_nlp/ocp_nlp_common.c [hunk 2]
```diff
@@ -3209,7 +3229,7 @@ void ocp_nlp_common_eval_lagr_grad_p(ocp_nlp_config *config, ocp_nlp_dims *dims,
     int N = dims->N;
     // int *nu = dims->nu;
     // int *nx = dims->nx;
-    int *np = dims->np;
+    int np_global = dims->np_global;
 
     if (!strcmp("params_stage", field))
```
Comment (sandmaennchen) [inline acados/ocp_nlp/ocp_nlp_common.c:3234]: rename to `p`?

File: interfaces/acados_template/acados_template/acados_ocp.py [hunk 3]
```diff
@@ -845,6 +847,8 @@ def make_consistent(self, is_mocp_phase=False) -> None:
                     raise Exception(f'with_solution_sens_wrt_params is only implemented if constraints depend not on parameters. Got parameter dependency for {horizon_type} constraint.')
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/acados_ocp.py:847]: "don't depend on"

File: interfaces/acados_template/acados_template/acados_ocp_solver.py [hunk 4]
```diff
@@ -510,7 +510,7 @@ def eval_and_get_optimal_value_gradient(self, with_respect_to: str = "initial_st
 
         Notes:
         - for field `initial_state`, the gradient is the Lagrange multiplier of the initial state constraint.
-        The gradient computation consist of adding the Lagrange multipliers correspondin to the upper and lower bound of the initial state.
+        The gradient computation consist of adding the Lagrange multipliers corresponding to the upper and lower bound of the initial state.
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/acados_ocp_solver.py:513]: "consists"

File: interfaces/acados_template/acados_template/acados_ocp_solver_pyx.pyx [hunk 5]
```diff
@@ -255,7 +255,7 @@ cdef class AcadosOcpSolverCython:
 
         Notes:
         - for field `initial_state`, the gradient is the Lagrange multiplier of the initial state constraint.
-        The gradient computation consist of adding the Lagrange multipliers correspondin to the upper and lower bound of the initial state.
+        The gradient computation consist of adding the Lagrange multipliers corresponding to the upper and lower bound of the initial state.
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/acados_ocp_solver_pyx.pyx:258]: "consists"

**Review by sandmaennchen — APPROVED** [review]
nice! :heart_hands:

**Comment by kisisjrlly** [comment]
Very impressive work. But I still have the question, why not implement sensitivity calculations for different stage parameter values p? Although the fact that the precomputation and parameter vectors are defined only once makes this implementation more efficient, this approach cannot satisfy all scenarios. For example, for some time-varying optimal control problems or scenarios where the parameter values of different stages are determined by other modules (such as the output of neural networks)

**Comment by FreyJo** [comment]
Hi @kisisjrlly 
It is possible to use the stage wise parameters to select different parts of `p_global` and basically implement sensitivities wrt stage wise varying parameters. We do not have an example with this yet, but already thought about it and do expect such a formulation to be very efficient, such that no additional implementation in C is needed.
Probably, such a formulation will become a feature of [leap-c](https://github.com/leap-c/leap-c) soon.

**Pull request merged**